Avoid merge conflicts in CHANGELOG.md

jcrist opened the pull request:
Instead of having every PR modify `CHANGELOG.md` themselves (which
easily led to merge conflicts), we now store changes in the `changes/`
directory. Each PR adds a single file to the directory describing the
change. On release, the `update_changelog.py` script can be run, which
compiles the changelog and updates `CHANGELOG.md` (and clears the
changes directory to restart this process). An `EXAMPLE.yaml` changelog
entry will persist between clears to provide some documentation of this
process.

Fixes #2311.

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

jcrist commented on changes/EXAMPLE.yaml:
```diff
@@ -0,0 +1,23 @@
+  - "Example change description - [#1234](https://github.com/PrefectHQ/prefect/pull/1234)"
+
+contributor:
+  - "[Contributor Name](https://github.com/contributor_github_username)"
```
We could definitely change this to be more structured if we wanted (for example, take in github issue/pr numbers and autogenerate the links). I went with the simplest thing possible for now.

jcrist replied:
Example with more structure:

```yaml
description: "Example change description"
kind: enhancement
links:
  - issue: 1234
  - pr: 1235
contributor:
  name: Contributor Name
  github: contributor_username
```

Multiple changelog entries in a single file (rare) could be separated by `---`, and loaded a yaml stream.

I'm not convinced this is better, it's just an option. The current approach might be easier for those familiar with markdown (it's also simpler to implement). The more structured approach might be easier for those that aren't, and could let us enforce changelog formatting more.

jcrist commented:
> My two cents is that (for me personally as a dev) creating and committing new files is on par with resolving a changelog merge conflict in the GitHub UI

I generally agree with this, but think that as we work more in parallel (and with more external contributors), the pain of the current strategy will get worse. I resolved a merge conflict 3 times a few weeks ago before the PR was finally merged, which was a wee bit frustrating. Note that the current changelog file on master is also malformed due to merge issues from the 0.11.0 release branch (there's two changelog headers, and two unreleased sections (which may be intentional?)). If a PR lapses between releases (which happen weekly), the PR has to update the section it is added to the changelog, or when it is merged it will be added to the incorrect release notes (I fixed one of these last week).

cicdw commented:
Yea, those are good points.  I'm +1

jlowin commented:
I'm a big fan - I hate changelog conflicts passionately. In addition to updating our development docs I have two suggestions:

1. Take the comment at the top of `EXAMPLE.yaml` and add it to a README in the `changes/` directory, purely to help users discover the instructions (I still really like the `EXAMPLE.yaml` file, I just was pleasantly surprised to find it contained instructions in addition to an example)

2. (perhaps not for this PR) this feels like it could be the beginning of a `prefect dev` CLI group. Something like `prefect dev create-changelog --pr 2402`

lauralorenz commented:
For posterity, @joshmeek myself and @jcrist discussed IRL yesterday, and we are for it! In response to https://github.com/PrefectHQ/prefect/pull/2529#discussion_r422417887, we agreed that simpler is better and to leave the metadata yaml in its current format. We felt good to merge and use it by hand for the next few releases, and then add it to automation on CircleCI somehow.

jlowin commented on update_changelog.py:
```diff
@@ -0,0 +1,131 @@
+import yaml
+
+SKIPLINES = 1
+SECTIONS = [
```
Just a note - I believe "server" was added as a section in this release
```suggestion
SECTIONS = [
    "server",
```

jcrist commented:
I've updated this with master, updated the developer docs, updated the PR template, and added a readme. To avoid having to convert existing entries over to yaml files, I think we should merge this right after the next release (maybe tuesday?).
